feat(vhd-lib): merge resume can resume when rename fails

fbeauchamp opened the pull request:
review by commit, do not squash 

### Check list

> Check if done, if not relevant leave unchecked.

- [ ] PR reference the relevant issue (e.g. `Fixes #007` or `See xoa-support#42`)
- [ ] if UI changes, a screenshot has been added to the PR
- [ ] documentation updated
- `CHANGELOG.unreleased.md`:
  - [x] enhancement/bug fix entry added
  - [x] list of packages to release updated (`${name} v${new version}`)
- **I have tested added/updated features** (and impacted code)
  - [x] unit tests (e.g. [`cron/parse.spec.js`](https://github.com/vatesfr/xen-orchestra/blob/b24400b21de1ebafa1099c56bac1de5c988d9202/%40xen-orchestra/cron/src/parse.spec.js))
  - [ ] if `xo-server` API changes, the corresponding test has been added to/updated on [`xo-server-test`](https://github.com/vatesfr/xen-orchestra/tree/master/packages/xo-server-test)
  - [x] at least manual testing

### Process

1. create a PR as soon as possible
1. mark it as `WiP:` (Work in Progress) if not ready to be merged
1. when you want a review, add a reviewer (and only one)
1. if necessary, update your PR, and re- add a reviewer

From [_the Four Agreements_](https://en.wikipedia.org/wiki/Don_Miguel_Ruiz#The_Four_Agreements):

1. Be impeccable with your word.
1. Don't take anything personally.
1. Don't make assumptions.
1. Always do your best.

julien-f commented on packages/vhd-lib/Vhd/VhdAbstract.js:
```diff
@@ -202,10 +202,17 @@ exports.VhdAbstract = class VhdAbstract {
+        return VhdAbstract.rename(handler, sourcePath, targetPath)
+      }
+      throw e
+    }
```
Do not support renaming `VhdDirectory`.

julien-f replied:
Maybe even remove this method and use `Handler#rename` directly.

julien-f commented on packages/vhd-lib/Vhd/VhdAbstract.js:
```diff
@@ -200,14 +200,6 @@ exports.VhdAbstract = class VhdAbstract {
-    } catch (e) {}
-    await handler.rename(sourcePath, targetPath)
-  }
-
```
IMHO, this fix should be in a separate commit, what do you think? :slightly_smiling_face:

fbeauchamp replied:
yes. I kept all the commit separated to eas th review, if the review is ok  , I will squash the first four commits, and rework the last one with a detailled explanation  ( I am sure I will berate past me in the future for removing this )

fbeauchamp committed:
fix(vhd-lib): remove unsafe VhdAbstract.rename implementation

actual implementation was deleting the target vhd even if the source did not exist, leading to ptential data loss

fbeauchamp committed:
feat(vhd-lib): implement merge  resuming during rename phase

fbeauchamp committed:
refactor

julien-f commented on packages/vhd-lib/merge.js:
```diff
@@ -41,162 +41,191 @@ const { warn } = createLogger('vhd-lib:merge')
+
+  constructor(handler,
+    chain,
+    { onProgress = noop, logInfo = noop, removeUnused = false, mergeBlockConcurrency = 2 } = {}){
```
Remove these default values, trust the calling code. :slightly_smiling_face:

julien-f commented on packages/vhd-lib/merge.js:
```diff
@@ -212,19 +241,70 @@ module.exports.mergeVhdChain = limitConcurrency(2)(async function mergeVhdChain(
+      }
+    })
+    // done
+    return this.#cleanup()
```
Not part of `cleanVhds` handles it in the `switch`.

julien-f commented on packages/vhd-lib/merge.js:
```diff
@@ -41,162 +41,191 @@ const { warn } = createLogger('vhd-lib:merge')
 
+  }
+
+  async * #prepare(){
```
Unclear name, `#openVhds` maybe?

julien-f commented on CHANGELOG.unreleased.md:
```diff
@@ -33,6 +33,7 @@
 
 <!--packages-start-->
 
+- vhd-lib minor
```
Duplicate, should be removed.

julien-f commented on packages/vhd-lib/merge.js:
```diff
@@ -41,162 +41,189 @@ const { warn } = createLogger('vhd-lib:merge')
+
+  constructor(handler,
+    chain,
+    { onProgress, logInfo, removeUnused , mergeBlockConcurrency = 2 } = {}){
```
Don't duplicate default values.

fbeauchamp committed:
fix(vhd-lib): merge cleanup data more thoroughly

fbeauchamp committed:
fix
